### ISSUE TITLE
Jax.random Very Slow in JIT compared to Pmap under multi-gpus

### ISSUE THREAD

**Issue by bzhangGo** [google/jax#14559]
### Description

I'm using JIT (Jax.Array) for data parallelization, but generating a random matrix as required in dropout layers is **very inefficient** in JIT under the multi-gpu setting.

Below is an example code:
```python
import functools
import numpy as np
import jax
import time

from jax import lax, random, numpy as jnp

from jax.experimental.pjit import pjit, with_sharding_constraint
from jax.sharding import Mesh, PartitionSpec, NamedSharding
from jax.experimental import mesh_utils, maps


# Start a device mesh.
device_mesh = mesh_utils.create_device_mesh((4, 1))
# Annotate each axis with a name.
mesh = Mesh(devices=device_mesh, axis_names=('data', 'model'))


def dropout_layer(x, rate, rng_key, share_dims=None):
  if isinstance(rate, float) and 0.0 < rate < 1.0: 
    keep_rate = 1. - rate

    if share_dims is not None:
      if not isinstance(share_dims, (list, tuple)):
        share_dims = [share_dims]
    else:
      share_dims = []

    broadcast_shape = list(x.shape)
    for dim in share_dims:
      broadcast_shape[dim] = 1
    mask = jax.random.uniform(rng_key, broadcast_shape, x.dtype) < keep_rate
    mask = jnp.broadcast_to(mask, x.shape)
    return jax.lax.select(mask, x / keep_rate, jnp.zeros_like(x))
  else:
    return x



# Create fake inputs.
x = jnp.ones((96, 256, 512))
# Initialize a PRNG key.
k = random.PRNGKey(0)


import sys

method = sys.argv[1]


def f(x, k):

  for _ in range(10):
    x = dropout_layer(x, 0.5, k)
  y = x

  if method == 'pjit':
    return y.sum()
  else:
    return jax.lax.psum(y.sum(), axis_name='data')



if method == "pjit":
  jit_f = jax.jit(f)

  with mesh:
    start_time = time.time()
    x = jax.device_put(x, NamedSharding(mesh, PartitionSpec('data')))
    y = jit_f(x, k)

    for _ in range(100):
      y = jit_f(x, k)
    print(time.time() - start_time)

else:
  pmap_f = jax.pmap(f, axis_name='data')
  x = x.reshape((4, -1) + x.shape[1:])
  k = jax.device_put_replicated(k, jax.local_devices())

  with mesh:
    start_time = time.time()
    y = pmap_f(x, k)

    for _ in range(100):
      y = pmap_f(x, k)
    print(time.time() - start_time)
```

Running with 4 GPUs gives the following timing results:
```
CUDA_VISIBLE_DEVICES=4,5,6,7 python script.py pjit
12.549862623214722
CUDA_VISIBLE_DEVICES=4,5,6,7 python script.py pmap
2.7941601276397705
```
I don't know why the speed difference is so huge. Could anyone give some explanation? If this is just my coding's problem, please point it out; otherwise, it seems that some operations are not well supported by jax.Array to efficiently handle data/model parallelization.

Thanks,

### What jax/jaxlib version are you using?

jax/jaxlib: 0.4.2

### Which accelerator(s) are you using?

4 GPU 

### Additional system info

python=3.8

**Comment by apaszke** [comment]
This is unfortunately expected. The PRNG JAX uses by default interacts badly with the automatic partitioning in XLA. We're working on rectifying that issue, but the fix is still in progress. In the meantime you could use the `jax_default_prng_impl` config to switch to one of the implementations that are more friendly to automatic partitioning. See [the relevant docs here](https://jax.readthedocs.io/en/latest/jax.random.html#advanced-rng-configuration).

**Comment by apaszke** [comment]
I found a better reference actually. There's [a flag `jax_threefry_partitionable`](https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html#generating-random-numbers) that might help you.

**Comment by bzhangGo** [comment]
Thanks, @apaszke. Yeah, adding this flag works!

**Comment by XilunWu** [comment]
I'm quite curious of what data is communicated among devices when the flag is False?

**Comment by XilunWu** [comment]
@mattjj Thx for the quick answer and sorry for bring back the unhappy memory ;-)

To summarize, the old impl has reshape and concatenate calls which introduce the extra communication, and the new impl doesn't have them but each device simply doing threefry hash locally on the `count` array (which is sharded). Do I understand correctly?

**Comment by froystig** [comment]
Yep, at a high enough level, that's correct. There are a few more details to accommodate different output bit-widths, and the fact that the count array must comprise 64-bit ints. But essentially we are now hashing something equivalent to `iota(n)` when generating `n` random words, and the automatic partitioner can clearly discover how to do this "locally" only.
